### PR TITLE
Section 2: Working with numbers

### PR DESCRIPTION
**Pull request by ziqiang2344**
- https://github.com/ziqiang2344/Learn_Ruby_on_Rails/pull/3 needs to be merged before this PR can be merged

### CODE REVIEW

File: sec2_the_ruby_programming_language/2.3_number.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# Numbers
+puts 1 + 2       #3
+puts 2 - 1       #1
+puts 10 * 2      #20
+puts 10 / 4      #2
+puts 10.0 / 4    #2.5
```
Comment (tlkiong) [inline sec2_the_ruby_programming_language/2.3_number.rb:6]: why these happen here?

File: sec2_the_ruby_programming_language/2.3_number.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+puts 10 / 4.0    #2.5
+puts 10 / 4.to_f #2.5
+puts (10/4).to_f #2.0
```
Comment (tlkiong) [inline sec2_the_ruby_programming_language/2.3_number.rb:9]: Why?

File: sec2_the_ruby_programming_language/2.3_number.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+# Assign variables
+x = 5
+y = 10
+puts y / x #2
```
Comment (tlkiong) [inline sec2_the_ruby_programming_language/2.3_number.rb:14]: Based on line 1-9, this formula will yield an incorrect result no?

File: sec2_the_ruby_programming_language/2.3_number.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+puts "hello".to_f #0.00
+
+#Simple calculator
+puts "Simple calculator"
+25.times { print "-"}
+puts "\nEnter the first number"
+num_1 = gets.chomp
+puts "Enter the second number"
+num_2 = gets.chomp
+puts "The first number multiplied by the second number is #{num_1.to_i * num_2.to_i}"
```
Comment (tlkiong) [inline sec2_the_ruby_programming_language/2.3_number.rb]: Why `.to_i`? If you key in:
```
num_1 = 2.5
num_2 = 3
```
it will be wrong answer

**Review by tlkiong — CHANGES_REQUESTED** [review]
Check comments.

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 1 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:6]: This is because the answer will follow the data type of the left operand of the arithmetic operator.

When the integer is divided by an integer, the end result will be an integer.
When the float number is divided by an integer, the end result will be a float number.

However, when both operands have the same data type then the end result will also have the same data type as both operands.

Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:6]: This is because many binary operators that expect operands of arithmetic type cause conversions and yield result types in a similar way. This pattern is called the usual arithmetic conversions. 

In this case, when the float number is divided by an integer, the end result will also be converted into a float number automatically.

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 2 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:9]: This is considered as the usual arithmetic conversions, when either type is the float number, the end result will also become a float number.

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 3 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:14]: Is the incorrect result referring to the conversion?

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 4 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:46]: Understood. Will remove it from the code.

Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb]: This is because the `num_1` and `num_2` variables are assigned to input with string type. Hence, we will need to convert the `num_1` and `num_2` implicitly into integer types so that the multiplication can be carried out successfully.

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 1 above)
Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb:6]: Speak english please

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 2 above)
Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb:9]: No idea how does that answer the question.

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 3 above)
Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb:14]: wat do you mean?

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 4 above)
Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb]: ok. So, you are telling me need to convert to to_i. I never refute that. I am saying by doing so, you will yield a wrong answer. So, no idea what you are talking about here

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 1 above)
Comment (Michelle-Lohwt) [reply sec2_the_ruby_programming_language/2.3_number.rb:6]: Just explain like you explain to your grandma will do😂 @ziqiang2344

Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:6]: 10 / 4 = 2 in line 5 because it is an integer where there is no decimal places tailing behind.
10.0 /4 = 2.5 in line 6 because it is a float number

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 2 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:9]: This is because all the answers are converted to become a float number.

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 3 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:14]: This will not yield an incorrect result.

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 1 above)
Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb:6]: I can see that. But I don't understand why. My question is why.
Logically with maths, 10/4 == 2.5. Why in this case it isn't?

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 2 above)
Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb:9]: Still have no idea what are you talking about.
Are you sure you know what exactly I am asking about?

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 4 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb]: Okay sure understood. Thank you.

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 3 above)
Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb:14]: Change x to 3. Do you get 3.33333 or 3?
The correct answer should be 3.333333333 right?

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 2 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:9]: I'm not sure about the question. May I have more clarification?

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 1 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:6]: Ya, the answer will not have any decimal places unless we convert it to become a float number. Else, it will become the answer will assigned the data type same as the operands.

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 3 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:14]: The answer will remain as 3 unless it is converted to become a float number.

Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb:14]: Hello. Please keep up with the conversation.
Please read back on the **entire** conversation and if don't understand what we are discussing, please ask

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 2 above)
Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb:9]: My question: 

```
(10/4).to_f
```

Think in terms of mathematics:
10 divide by 4 is 2.5.

so:
(10/4) = (2.5)
(2.5).to_f = 2.5

Why is it 2.0 instead of 2.5?

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 1 above)
Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb:6]: > the answer will not have any decimal places unless we convert it to become a float number

Based on what you say, unless we convert `it` to become a float number. Thats saying it should be correct if we convert the answer into decimal place right? Which is [this](https://github.com/ziqiang2344/Learn_Ruby_on_Rails/pull/4#discussion_r846716080). But, this seems to show differently.

I really confused with your explanation

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 2 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:9]: This is because the computation will be carried out according to the precedence level of the operator. Based on the picture here, the parentheses have higher precedence over another operator.

In this case, the `10/4` is being encapsulated in the parenthesis in which it will be computed first then only will be converted into a float number by `.to_f`

The sequence of the computation shall be looked like as below:
1. `(10/4).to_f`
2. `(2).to_f`
3. `2.0` (final answer)

image

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 1 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:6]: So the question here is asking why 10/4 is equal to 2 instead of 2.5. 

In this case, 10 & 4 are both integer numbers, which also means that, after the computation, `10/4` will be an integer by default also. Hence, the answer will be a `2`

However, if we put convert one or both of the operands into a float number. For example, we convert `10` into `10.0` / `4` into `4.0` / 10 & 4 into `10.0` and `4.0`. Then, all of it will result in `2.5`.

In summary,
1. integer / integer = integer
2. integer /float = float
3. float/integer = float
4. float/float  = float

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 3 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:14]: So the question here is asking if we assign `3` to `x`, after `10/3`, why the answer will become `3` instead of `3.3333`. This is due to the reason mentioned [above](https://github.com/ziqiang2344/Learn_Ruby_on_Rails/pull/4#discussion_r851601181).

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 1 above)
Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb:6]: Ok. But, what I have been asking from the beginning is why `integer / integer = integer`

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 4 above)
Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb]: ??

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 1 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb:6]: This is because it is being specified in the [Ruby spec](https://github.com/ruby/spec/blob/master/core/integer/divide_spec.rb#L19-L21) that when an integer is divided by an integer, an integer division will occur and hence it will return an integer answer as well.

File: sec2_the_ruby_programming_language/2.3_number.rb (hunk 4 above)
Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb]: Is the wrong answer here referring to 2.5 * 2 supposed to be 5 but it will result in 4 instead?

Comment (tlkiong) [reply sec2_the_ruby_programming_language/2.3_number.rb]: huh?
no.
See my inputs above pls

Comment (ziqiang2344) [reply sec2_the_ruby_programming_language/2.3_number.rb]: I see the problem when 2.5 * 3, the output given here is 6 instead of 7.5. I'll remove this code section.

**Review by tlkiong — APPROVED** [review]
